src/test/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/resource/UserAuthenticationIntTest.kt

```diff
@@ -0,0 +1,260 @@
+        }
+      }
+      @AfterEach
+      internal fun deleteUsers() = dataBuilder.deleteAllUsers()
+
+      @Test
+      internal fun `will return service unavailable response`() {
+        webTestClient.post().uri("/authorise/reporting")
+          .headers(setAuthorisation(user = "marco.rossi"))
+          .exchange()
+          .expectStatus().isEqualTo(SERVICE_UNAVAILABLE)
+      }
+
+      @Test
+      internal fun `will return service unavailable for any error`() {
+        reportingAuthenticationApi.stubError(status = INTERNAL_SERVER_ERROR.value())
+        webTestClient.post().uri("/authorise/reporting")
+          .headers(setAuthorisation(user = "marco.rossi"))
+          .exchange()
+          .expectStatus().isEqualTo(SERVICE_UNAVAILABLE)
+      }
+    }
+
+    @Nested
+    @DisplayName("when the reporting service is available")
+    inner class ServiceIsAvailable {
+      @BeforeEach
+      internal fun setUp() {
+        reportingAuthenticationApi.stubSuccessfulAuthentication("https://reporting.justice.gov.uk/nomis?session=12345")
+      }
+
+      @AfterEach
+      internal fun deleteUsers() = dataBuilder.deleteAllUsers()
+
+      @Nested
+      @DisplayName("when user has 970 (Business Objects Power User) role on active caseload")
+      inner class BusinessObjectsPowerUserActiveCaseload {
+        @BeforeEach
+        internal fun createUsers() {
+          with(dataBuilder) {
+            generalUser()
+              .username("marco.rossi")
+              .atPrison("WWI")
+              .nomisRoles(listOf("970"))
+              .buildAndSave()
+          }
+        }
+
+        @Test
+        internal fun `will pass username to reporting service`() {
+          webTestClient.post().uri("/authorise/reporting")
+            .headers(setAuthorisation(user = "marco.rossi"))
+            .exchange()
+            .expectStatus().isEqualTo(OK)
+
+          reportingAuthenticationApi.verifyAuthenticationRequest(username = "marco.rossi")
+        }
+
+        @Test
+        internal fun `will retrieve a report url`() {
+          webTestClient.post().uri("/authorise/reporting")
+            .headers(setAuthorisation(user = "marco.rossi"))
+            .exchange()
+            .expectStatus().isEqualTo(OK)
+            .expectBody()
+            .jsonPath("$.reportingUrl").isEqualTo("https://reporting.justice.gov.uk/nomis?session=12345")
+        }
+      }
+
+      @Nested
+      @DisplayName("when user has 970 (Business Objects Power User) role on non active caseload")
+      inner class BusinessObjectsPowerUserNonActiveCaseload {
+        @BeforeEach
+        internal fun createUsers() {
+          with(dataBuilder) {
+            generalUser()
+              .username("marco.rossi")
+              .atPrisons("WWI", "BXI")
+              .nomisRoles("200", "970")
+              .build()
+              .apply {
+                this.userPersonDetail.caseloads.removeRole(role = "970", prison = "WWI")
+                this.save()
```

Comment:
By default the builder will add the same roles to each caseload so need to hack the data to remove the role before it is saved